### ISSUE TITLE
CSS selectors ignore root node

### ISSUE THREAD

**Issue by rushter** [lexbor/lexbor#122]
In Modest, the `modest_finder_by_selectors_list` includes the specified node when searching for HTML.
In Lexbor, the first node (root node) gets ignored.

I use the following function:
`lxb_selectors_find(selectors, node, selectors_list, <lxb_selectors_cb_f>css_finder_callback, <void*>self)`

For example:
```
<div id="container">
    <span class="red">Hi</span>
</div>
```

If I pass the `span` node as a second argument to `lxb_selectors_find` the `.red` CSS query will return no results.

To reproduce, please change a few lines in the example:
https://github.com/lexbor/lexbor/blob/353b772429b9ec3ffb3194dedc4ddaecd7f31109/examples/lexbor/selectors/easy_way.c#L47-L46

```
    static const lxb_char_t html[] = "<body id='test'><div><p class='x z'></p><p id='y'></p></div></body>";

    /* CSS Data. */

    static const lxb_char_t slctrs[] = "#test";
```
The example uses `lxb_html_document_body_element` to get the body element and the CSS query will fail.

**Comment by lexborisov** [comment]
@rushter 

The logic is exactly this: where to look, inside what? Also `querySelectorAll` works in browsers.
For example, you will find the necessary nodes using one selector and want to find others for these nodes using the next selector.

Chrome example:
```HTML
<div>a</div>
<script>
    var node = document.body.querySelectorAll("body");
    console.log(node); /* empty */
</script>
```

```HTML
<div>a</div>
<script>
    var node = document.querySelectorAll("body");
    console.log(node); /* NodeList [body] */
</script>
```

**Comment by rushter** [comment]
> @rushter
> 
> The logic is exactly this: where to look, inside what? Also `querySelectorAll` works in browsers.
> For example, you will find the necessary nodes using one selector and want to find others for these nodes using the next selector.
> 
> Chrome example:
> 
> ```
> <div>a</div>
> <script>
>     var node = document.body.querySelectorAll("body");
>     console.log(node); /* empty */
> </script>
> ```
> 
> ```
> <div>a</div>
> <script>
>     var node = document.querySelectorAll("body");
>     console.log(node); /* NodeList [body] */
> </script>
> ```

What's is the best option to start from the same node? I think it's possible that a given node will have no parent, so starting from a parent node of a given node may not work for all the cases.

I need this to support chaining (which works as expected in Modest and I need to replicate it).

```python
html = """
<div id="container">
    <span class="red"><div></div></span>
    <span class="green"></span>
    <span class="red"></span>
    <span class="green"></span>
</div>
"""
tree = selectolax.lexbor.LexborHTMLParser(html)
tree.css('div').css("span").css(".red").matches
```

**Comment by lexborisov** [comment]
@rushter 

[...]
> I need this to support chaining (which works as expected in Modest and I need to replicate it).
> 
> ```python
> html = """
> <div id="container">
>     <span class="red"><div></div></span>
>     <span class="green"></span>
>     <span class="red"></span>
>     <span class="green"></span>
> </div>
> """
> tree = selectolax.lexbor.LexborHTMLParser(html)
> tree.css('div').css("span").css(".red").matches == 2
> ```

In the above chain, you have a logical error.

```Python
tree.css('div').css("span")
```
Here you are looking for a `div` with a `span` inside. Ok, this is correct, logical.

```Python
.css("span").css(".red")
```
But in this case, you already want to check the current (found) element against the selector.
One method performs different logic.

In the CSS it would look like this:
```CSS
div span.red
```
and:
```CSS
div span .red
```

You want your example to work in both cases. This is not right. And Modest does not work correctly in this place.
In other words, the user will not know what result he will get using the current approach with your example: 
```Python
tree.css('div').css("span").css(".red")
```

The only valid behavior in your example is this like a selector: 
```CSS
div span .red
```

**Comment by rushter** [comment]
> @rushter
> 
> [...]
> 
> > I need this to support chaining (which works as expected in Modest and I need to replicate it).
> > ```python
> > html = """
> > <div id="container">
> >     <span class="red"><div></div></span>
> >     <span class="green"></span>
> >     <span class="red"></span>
> >     <span class="green"></span>
> > </div>
> > """
> > tree = selectolax.lexbor.LexborHTMLParser(html)
> > tree.css('div').css("span").css(".red").matches == 2
> > ```
> 
> In the above chain, you have a logical error.
> 
> ```python
> tree.css('div').css("span")
> ```
> 
> Here you are looking for a `div` with a `span` inside. Ok, this is correct, logical.
> 
> ```python
> .css("span").css(".red")
> ```
> 
> But in this case, you already want to check the current (found) element against the selector.
> One method performs different logic.
> 
> In the CSS it would look like this:
> 
> ```css
> div span.red
> ```
> 
> and:
> 
> ```css
> div span .red
> ```
> 
> You want your example to work in both cases. This is not right. And Modest does not work correctly in this place.
> In other words, the user will not know what result he will get using the current approach with your example:
> 
> ```python
> tree.css('div').css("span").css(".red")
> ```
> 
> The only valid behavior in your example is this like a selector:
> 
> ```css
> div span .red
> ```

I know that this may look like a logical error, but it's designed this way.
Every next `css` call gets evaluated against all previous matches including the provided node.

The chaining feature supports extra methods and mimics some of the xpath features, this is why my example uses multiple `css` calls instead of one.
For example, it supports `select('script').text_contains('var counter = ').css("#counter")`

**Comment by rushter** [comment]
I think that's the only thing left, all other tests pass for now and both backends return the same data in simple cases.
I can't use `.parent` here, because that includes other parts of data as well.

**Comment by lexborisov** [comment]
@rushter 

You cannot use `.parent` here, you are right. 
But the current approach is also wrong. It can give false results and mislead users.

I can add `lxb_selectors_find_close()` function, but I would recommend reconsidering your approach.

**Comment by rushter** [comment]
> @rushter
> 
> You cannot use `.parent` here, you are right.
> But the current approach is also wrong. It can give false results and mislead users.
> 
> I can add `lxb_selectors_find_close()` function, but I would recommend reconsidering your approach.

I don't think there are better options, this is not the default selector which most users use.
I will disable the chained CSS part for lexbor for now. I don't use chained CSS calls myself, but I use extra features such as `text_contains`.

**Comment by rushter** [comment]
@lexborisov Do you have any idea on why this happens?

```
  hash.c
    lexbor/source\lexbor\core\hash.c(38): error C2491: 'lexbor_hash_insert_raw': definition of dllimport data not allowed
    lexbor/source\lexbor\core\hash.c(41): error C2491: 'lexbor_hash_insert_lower': definition of dllimport data not allowed
    lexbor/source\lexbor\core\hash.c(44): error C2491: 'lexbor_hash_insert_upper': definition of dllimport data not allowed
    lexbor/source\lexbor\core\hash.c(63): error C2491: 'lexbor_hash_search_raw': definition of dllimport data not allowed
    lexbor/source\lexbor\core\hash.c(66): error C2491: 'lexbor_hash_search_lower': definition of dllimport data not allowed
    lexbor/source\lexbor\core\hash.c(69): error C2491: 'lexbor_hash_search_upper': definition of dllimport data not allowed
    lexbor/source\lexbor\core\hash.c(115): warning C4273: 'lexbor_hash_create': inconsistent dll linkage
```

Full log: https://github.com/rushter/selectolax/runs/3393661157?check_suite_focus=true

I tried to build Windows version on selectolax and got this error.

**Comment by vtorri** [comment]
because LEXBOR_SHARED should be correctly defined. It's not currently the case

**Comment by rushter** [comment]
> because LEXBOR_SHARED should be correctly defined. It's not currently the case

Thank you for your answer, but I'm building statically. I got the idea of what I need to search for.

**Comment by vtorri** [comment]
core/def.h. If you build a static lexbor lib, then pass LEXBOR_STATIC to the preprocessor

... but on Windows, DLL are good (remark of a windows gcc dev)

**Comment by rushter** [comment]
> core/def.h. If you build a static lexbor lib, then pass LEXBOR_STATIC to the preprocessor
> 
> ... but on Windows, DLL are good (remark of a windows gcc dev)

I don't think I can use a shared library for my Python wrapper.
The users would need to install it first. Right now, they don't need to do anything. For modern versions of Windows, they just get a precompiled version of the library from Python Package Index.

**Comment by rushter** [comment]
Fixed the problem by setting `_WIN64` and `_WIN32` defines.

Are there any other important flags that I'm missing?

```python

        args = [
            "-pedantic", "-fPIC",
            "-Wno-unused-variable",
            "-Wno-unused-function",
            "-std=c99",
            "-O2",
            "-DLEXBOR_STATIC",
        ]
        win_args = [
            '-D_WIN64' if ARCH == '64bit' else '-D_WIN32',
        ]
```

**Comment by vtorri** [comment]
_WIN32 is always defined with a C Windows compiler, and _WIN64 is always defined for a 64 bits compiler. So pass -D_WIN32 unconditionnally on Windows. But if you use gcc, you should not be adding those flags (gcc already defines them correctly

-fPIC is for non Windows compilers, but it's for shared libs, not static libs
